pkg/proxy/util/nfacct/nfacct.go

```diff
@@ -0,0 +1,328 @@
+type runner struct {
+	handler handler
+}
+
+// New returns a new Interface.
+func New() (Interface, error) {
+	hndlr, err := newNetlinkHandler()
+	if err != nil {
+		return nil, err
+	}
+	return newInternal(hndlr)
+}
+
+// newInternal returns a new Interface with the given handler.
+func newInternal(hndlr handler) (Interface, error) {
+	return &runner{handler: hndlr}, nil
+
+}
+
+// Ensure is part of the interface.
+func (r *runner) Ensure(name string) error {
+	counter, err := r.Get(name)
+	if counter != nil {
+		return nil
+	}
+
+	if err != nil && errors.Is(err, ErrObjectNotFound) {
+		return handleError(r.Add(name))
+	} else if err != nil {
+		return handleError(err)
+	} else {
+		return ErrUnexpected
+	}
+}
+
+// Add is part of the interface.
+func (r *runner) Add(name string) error {
+	if name == "" {
+		return ErrEmptyName
+	}
+	if len(name) > MaxLength {
+		return ErrNameExceedsMaxLength
+	}
+
+	req := r.handler.newRequest(cmdNew, unix.NLM_F_REQUEST|unix.NLM_F_CREATE|unix.NLM_F_ACK)
+	req.AddData(nl.NewRtAttr(attrName, nl.ZeroTerminated(name)))
+	_, err := req.Execute(unix.NETLINK_NETFILTER, 0)
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// Get is part of the interface.
+func (r *runner) Get(name string) (*Counter, error) {
+	if len(name) > MaxLength {
+		return nil, ErrNameExceedsMaxLength
+	}
+
+	req := r.handler.newRequest(cmdGet, unix.NLM_F_REQUEST|unix.NLM_F_ACK)
+	req.AddData(nl.NewRtAttr(attrName, nl.ZeroTerminated(name)))
+	msgs, err := req.Execute(unix.NETLINK_NETFILTER, 0)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	var counter *Counter
+	for _, msg := range msgs {
+		counter, err = decode(msg, true)
+		if err != nil {
+			return nil, handleError(err)
+		}
+	}
+	return counter, nil
+}
+
+// List is part of the interface.
+func (r *runner) List() ([]*Counter, error) {
+	req := r.handler.newRequest(cmdGet, unix.NLM_F_REQUEST|unix.NLM_F_DUMP)
+	msgs, err := req.Execute(unix.NETLINK_NETFILTER, 0)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	counters := make([]*Counter, 0)
+	for _, msg := range msgs {
+		counter, err := decode(msg, true)
+		if err != nil {
+			return nil, handleError(err)
+		}
+		counters = append(counters, counter)
+	}
+	return counters, nil
+}
+
+var ErrObjectNotFound = errors.New("object not found")
+var ErrObjectAlreadyExists = errors.New("object already exists")
+var ErrNameExceedsMaxLength = fmt.Errorf("object name exceeds the maximum allowed length of %d characters", MaxLength)
+var ErrEmptyName = errors.New("object name cannot be empty")
+var ErrUnexpected = errors.New("unexpected error")
+
+func handleError(err error) error {
+	switch {
+	case err == nil:
+		return nil
+	case errors.Is(err, syscall.ENOENT):
+		return ErrObjectNotFound
+	case errors.Is(err, syscall.EBUSY):
+		return ErrObjectAlreadyExists
+	default:
+		return fmt.Errorf("%s: %s", ErrUnexpected.Error(), err.Error())
+	}
+}
```

Comment:
why is better to create new errors types than bubbling them up or just wrap them?

Reply:
It will be easier for callers to match on error, no?

Reply:
but this is forcing everybody to use your sentinels, I think you should wrap the existing errors so people can still assert on the original error, and only create new ones for the errors that does not exist

Reply:
fair enough